src/librustdoc/clean/inline.rs

```diff
@@ -546,22 +546,47 @@ fn build_static(cx: &mut DocContext<'_>, did: DefId, mutable: bool) -> clean::St
 fn build_macro(cx: &mut DocContext<'_>, did: DefId, name: Symbol) -> clean::ItemKind {
     let imported_from = cx.tcx.crate_name(did.krate);
     match cx.enter_resolver(|r| r.cstore().load_macro_untracked(did, cx.sess())) {
-        LoadedMacro::MacroDef(def, _) => {
-            let matchers: Vec<Span> = if let ast::ItemKind::MacroDef(ref def) = def.kind {
-                let tts: Vec<_> = def.body.inner_tokens().into_trees().collect();
-                tts.chunks(4).map(|arm| arm[0].span()).collect()
+        LoadedMacro::MacroDef(item_def, _) => {
+            let def = if let ast::ItemKind::MacroDef(ref def) = item_def.kind {
+                def
             } else {
                 unreachable!()
             };
-
-            let source = format!(
-                "macro_rules! {} {{\n{}}}",
-                name.clean(cx),
-                matchers
-                    .iter()
-                    .map(|span| { format!("    {} => {{ ... }};\n", span.to_src(cx)) })
-                    .collect::<String>()
-            );
+            let tts: Vec<_> = def.body.inner_tokens().into_trees().collect();
+            let matchers: Vec<Span> = tts.chunks(4).map(|arm| arm[0].span()).collect();
+
+            let name = name.clean(cx);
+            let source = if def.macro_rules {
```

Comment:
This code is mostly the same as the one in [clean/mod.rs](https://github.com/rust-lang/rust/blob/701496384fd2d46557b3f013b50860c69c9848cc/src/librustdoc/clean/mod.rs#L2177-L2208). Do we want to unify both or not? I'd tend to say yes but considering of the small differences (manipulating ast vs hir), not sure if the result would look better. So I thought about using a macro instead. What do you think?